### ISSUE TITLE
Tests

### ISSUE THREAD

**Issue by JelleVos1** [Wase-Engine/wase-engine#38]
**Is your feature request related to a problem? Please describe.**
When you work on a feature you might break another one without noticing. 

**Describe the solution you'd like**
Tests that test certain features and/or methods that run when new code is being merged

**Comment by requizm** [comment]
This is really good and important issue. Before we begin, we need to decide exactly what to do. Writing tests in games has been more difficult than in other software areas. Here comes some questions:  
  
  
**1. What tests will we do?**  
-Unit Test: I couldn't find a suitable unit test for this project. Because everything is interconnected. (scene-entity-component)
-Integration Test: Collision test(collision of two colliders), ui tests(button click test), entity tests(Add, delete), component tests(add, delete)
-Automated Test: Similar to integration test but this is more detailed. Maybe we can make small simulations for test. An example, open game window then simulate some clicks. But it looks so complicated. Amogus 😳
  
  
**2. Which library will we use?**  
[GoogleTest](https://github.com/google/googletest)? Or will we write it ourselves? We can use 3rd library for unit tests and integration tests. But i don't know about automated test.

**Comment by JelleVos1** [comment]
> This is really good and important issue. Before we begin, we need to decide exactly what to do. Writing tests in games has been more difficult than in other software areas. Here comes some questions:
> 
> **1. What tests will we do?**
> -Unit Test: I couldn't find a suitable unit test for this project. Because everything is interconnected. (scene-entity-component)
> -Integration Test: Collision test(collision of two colliders), ui tests(button click test), entity tests(Add, delete), component tests(add, delete)
> -Automated Test: Similar to integration test but this is more detailed. Maybe we can make small simulations for test. An example, open game window then simulate some clicks. But it looks so complicated. Amogus 😳
> 
> **2. Which library will we use?**
> [GoogleTest](https://github.com/google/googletest)? Or will we write it ourselves? We can use 3rd library for unit tests and integration tests. But i don't know about automated test.

I should start of by saying that I have never done testing before in C++ so it is a bit new for me so I can't really give a good opinion on this. I think like you say testing adding components and all these methods is a good place to start making tests on. I guess Google Tests would be nice, I read a lot of good things about it but I have never implemented it myself so I don't know if it is good for this use case, I hope you can clear that up for me. Also how do we link Google Tests? Does every developer need to install Google Tests like they have to do now with SDL or isn't that the case?

**Comment by requizm** [comment]
@JelleVos1 
I have never done testing before in C++ too. But it shouldn't be hard. 

Note: I misspelled above. We can only do unit testing with Google Test. 
  
**Linking Google Test**  
basic setup and testing: https://google.github.io/googletest/quickstart-cmake.html
<details>
  <summary>Example cmakefile of basic testing</summary>

```cmake
cmake_minimum_required(VERSION 3.17)
project(wase-engine)
# other wase-engine stuffs
# other wase-engine stuffs
# other wase-engine stuffs
# other wase-engine stuffs

include(FetchContent)
FetchContent_Declare(
  googletest
  URL https://github.com/google/googletest/archive/609281088cfefc76f9d0ce82e1ff6c30cc3591e5.zip
)
# For Windows: Prevent overriding the parent project's compiler/linker settings
set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
FetchContent_MakeAvailable(googletest)

enable_testing()

add_executable(
  button_test
  button_test.cc
)
target_link_libraries(
  button_test
  gtest_main
)

include(GoogleTest)
gtest_discover_tests(button_test)
```
</details>

More detailed testing(we will probably use this): https://google.github.io/googletest/primer.html

TLDR; I'm going to deal with another issue right now, once it's done I can start dealing with the test.

**Comment by JelleVos1** [comment]
> @JelleVos1
> I have never done testing before in C++ too. But it shouldn't be hard.
> 
> **Linking Google Test**
> Basic setup and testing: https://google.github.io/googletest/quickstart-cmake.html
> 
> Example cmakefile of basic testing
> More detailed testing(we will probably use this): https://google.github.io/googletest/primer.html
> 
> TLDR; I'm going to deal with another issue right now, once it's done I can start dealing with the test.

Cool, I will research more about the testing tonight and tomorrow too so I hope to get a bit more knowledge about it then

**Comment by requizm** [comment]
Now that the CI stuff is done, we can focus on this issue.

**Comment by JelleVos1** [comment]
> 
> 
> Now that the CI stuff is done, we can focus on this issue.

I looked into Google Tests a bit but I couldn't really figure it out. Maybe making a custom test system is better. What is your opinion about this?

**Comment by requizm** [comment]
@JelleVos1 
We can make a custom test system for automated tests. But wouldn't it be better to use a library like GoogleTest for simple functions to work and give the expected result? 

I made a basic example for this project:
https://github.com/requizm/wase-engine/commit/8788b220305062d156dccec5e058e0feec43bbc9

This is basic Unit Test. 

Test output:
```
[ctest]     Start 3: Scene.RemoveComponent
[ctest]     Start 2: Scene.GetComponent
[ctest]     Start 1: Scene.AddEntity
[ctest] 1/3 Test #2: Scene.GetComponent ...............   Passed    0.10 sec
[ctest] 2/3 Test #3: Scene.RemoveComponent ............   Passed    0.12 sec
[ctest] 3/3 Test #1: Scene.AddEntity ..................   Passed    0.12 sec
[ctest] 
[ctest] 100% tests passed, 0 tests failed out of 3
[ctest] 
[ctest] Total Test time (real) =   0.15 sec
[ctest] CTest finished with return code 0
```


Note: Don't worry about github action fails. Windows machines gives fail because test.exe needs to have dll files in the running location. We'll fix that bug later.

**Comment by JelleVos1** [comment]
> 
> 
> @JelleVos1
> We can make a custom test system for automated tests. But wouldn't it be better to use a library like GoogleTest for simple functions to work and give the expected result?
> 
> I made a basic example for this project:
> [requizm@35ff5e1](https://github.com/requizm/wase-engine/commit/35ff5e141206b113dd3259e5e16dc0635d6bb1dc)
> 
> This is basic Unit Test.
> 
> Test output:
> 
> ```
>     Start 1: BasicTests.AddEntity
> 1/3 Test #1: BasicTests.AddEntity .............   Passed    0.02 sec
>     Start 2: BasicTests.GetComponent
> 2/3 Test #2: BasicTests.GetComponent ..........   Passed    0.01 sec
>     Start 3: BasicTests.RemoveComponent
> 3/3 Test #3: BasicTests.RemoveComponent .......   Passed    0.01 sec
> 
> 100% tests passed, 0 tests failed out of 3
> 
> Total Test time (real) =   0.11 sec
> ```

This seems good! We will just go with Google Test then.

**Comment by requizm** [comment]
@JelleVos1 
So, i'm starting to write more regular unit tests.
